[jit][script] mutable lists

suo opened the pull request:
This PR implements the design that we discussed. Changes:
- Added a World token IValue and type. The IValue is basically a dummy struct for now, in the future we may extend it (say, add thread-local state).
- Effectful ops explicitly declare they are mutable by having World tokens as inputs and outputs in their schema.
- Purely functional ops that use mutable values will get "fenced" and the world token will be threaded through the fences
- AnnotateEffects pass which wires up all the world tokens together.

suo committed:
mutable lists, take 2

michaelsuo commented on torch/csrc/jit/passes/constant_propagation.cpp:
```diff
@@ -115,9 +118,10 @@ bool removeExtraNodeOutputs(Node *n) {
-    });
+      std::all_of(n->inputs().begin(), n->inputs().end(), [&](Value* v) {
+                           return v->node()->kind() == prim::Constant &&
+                               v->node()->output()->type() != WorldType::get();
```
This is gross, but we only need this until we can pass World tokens in and out of graphs

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -486,19 +517,26 @@ at::optional<std::vector<Value*>> tryMatchSchema(
-  Symbol name,
-  at::ArrayRef<NamedValue> inputs,
-  at::ArrayRef<NamedValue> attributes) {
+    const std::shared_ptr<Operator>& op,
```
tryEmitBuiltin is used by `graph->insert` when inserting nodes as well. So this modification will let _any_ pass put a node in by making up a world token. This is unsafe since passes that happen after the compiler are not going to call annotate effects. Furthermore, it _prevents_ those passes from actually passing a valid world token when creating an op. We should probably distinguish between calls within compiler.cpp which do implicitly insert the world token and calls from graph->insert which should not.

michaelsuo replied:
Makes sense. I'll lift these changes out into a `tryEmitMutableBuiltin` function and just rely on callers to check whether the schema is mutable and dispatch to the right emit function if so.

zdevito commented on torch/csrc/jit/passes/annotate_effects.cpp:
```diff
@@ -0,0 +1,227 @@
+    // and output of the graph. That would require changing a bunch of interop
+    // and batching code, so leaving it out for now.
+    //
+    // auto curToken = g->addInput("world")->setType(WorldType::get());
```
Discussed in person: we can add `g->entryWorld()` and `g->exitWorld()` to the graph. These would not be members of `inputs()` or `outputs()` of the graph but come from separate "special" nodes (which, similar to prim::Params, are not in the node list) that are used only to capture effects.

zdevito commented on torch/csrc/jit/passes/annotate_effects.cpp:
```diff
@@ -0,0 +1,227 @@
+  }
+
+  bool isMutatingOp(const Node* node) {
+    return std::any_of(
```
Shouldn't this be: `return node->inputs().size() > 0 && node->inputs()[0]->type() == WorldType::get()`. When you replace the world node you only ever replace the first entry.

zdevito commented on torch/csrc/jit/passes/annotate_effects.cpp:
```diff
@@ -0,0 +1,227 @@
+  bool shouldAnnotate(Node* node) {
+    // Check if this node uses a known mutable value
+    for (auto* input : node->inputs()) {
+      if (input->type()->kind() != TypeKind::ListType) {
```
Maybe best to encapsulate this now as: `if(!isMutableType(input->type()))`. In the future I think we will likely keep the set of mutable types and immutable types separate.

zdevito commented on torch/csrc/jit/passes/annotate_effects.cpp:
```diff
@@ -0,0 +1,227 @@
+    return curToken;
+  }
+
+  bool shouldAnnotate(Node* node) {
```
Why cache mutableValues_, but not shouldAnnotate on nodes/blocks?

zdevito commented on torch/csrc/jit/passes/annotate_effects.cpp:
```diff
@@ -0,0 +1,227 @@
+        node->owningGraph()->create(prim::MemoryFence, /*outputs=*/2);
+
+    endFence->addInput(curToken);
+    endFence->addInput(node->outputs().at(0));
```
Also here there is no guarantee that there is only one output of the original node. All values should go through the fence.

zdevito replied:
We should also re-wire all uses of the outputs to be the outputs after then fence, in order to prevent unexpected cases where other ops can get scheduled inside the fence operator.

zdevito replied:
Please fence _all_ inputs and _all_ outputs. The reason for doing this is to maintain fiction that we really just have two versions of every op: one that takes control dependencies and one that doesn't. Fencing only some of the inputs and outputs allows some re-writes that break that invariant.  At the very least it will make it harder to reason about what might happen. Later when we make mutability a property of the type, we can be much looser about when memory fences get inserted and how world tokens are threaded. (e.g. we can allow reads of mutable values to happen in any order and only serialize writes).

zdevito reviewed (commented):
This looks good -- I have a bunch of minor issues that should be fixed before we merge it, but nothing that should require major changes.

michaelsuo commented on torch/csrc/jit/function_schema.h:
```diff
@@ -69,13 +74,42 @@ struct FunctionSchema {
   const bool is_varret;
+  const bool is_mutable;
+
+  std::vector<Argument> getArgumentsMutable() const {
```
Whoops, left over from a previous attempt. Will remove…

michaelsuo commented on torch/csrc/jit/ir.h:
```diff
@@ -244,6 +244,17 @@ struct Value {
+  // Result:  %3 = f(%1, %2)
+  //          %4 = g(%5)
+  //          %5 = h(%4)
+  void removeAllUses();
```
ditto above, I'll remove

michaelsuo commented on torch/csrc/jit/ivalue.h:
```diff
@@ -101,13 +101,18 @@ struct ConstantString : at::Retainable {
 
+struct World {
+  World() : world_id(1) {}
+  int64_t world_id;
```
Nothing, it's just a placeholder, since empty structs are not allowed. In the future, we could put basic debugging info in there, or thread local state, but I couldn't thinking of anything useful so far.

apaszke commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -454,10 +473,14 @@ Operation listSlice(Node* node) {
 
 RegisterOperators reg2({
+    // Select element in the `b`th position from list `a`
+    // Equivalent to `a[b]` in Python.
```
Let's just please call the generic type `any` and not `Object` 😊

apaszke commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -394,6 +395,36 @@ static inline bool isIntUsedAsIntList(
+    std::ostream& failure_messages) {
+  JIT_ASSERT(schema.is_mutable);
+  // Add a dummy world token to be matched against
+  const auto worldToken = graph.insertConstant(World());
```
Can we verify that all those tokens are gone after annotating the effects?

apaszke replied:
We could probably add a check in `Graph::lint()` that makes sure that there's only a single world token constant

apaszke reviewed (commented):
I think there might be some extra changes required to make function inlining work correctly. You should remove the inline function's world constant, and use the current value of the caller's world. Similarly, the caller's world should be replaced with whatever is the final value in the callee.

michaelsuo commented on torch/csrc/jit/passes/annotate_effects.cpp:
```diff
@@ -0,0 +1,227 @@
+    startFence->insertBefore(node);
+
+    // modify the node to take in the start fence's output list
+    node->replaceInput(0, mut);
```
Good point. I'll add some notes to places that assume that we're dealing with a mutable list. Maybe it's worth adding an assert that we're annotating whitelisted ops, at least until we can make the function schema express which argument is getting mutated.

michaelsuo commented on torch/csrc/jit/passes/annotate_effects.cpp:
```diff
@@ -0,0 +1,227 @@
+
+    // For mutating ops, just thread the world token through the node.
+    if (isMutatingOp(node)) {
+      node->replaceInput(0, curToken);
```
It uses a "dummy" token generated by the compiler when emitting a mutable node. I'll add a comment.

michaelsuo commented on torch/csrc/jit/passes/annotate_effects.cpp:
```diff
@@ -0,0 +1,227 @@
+  //
+  // Returns the last world token emitted for subsequent annotations to use.
+  Value* visitNode(Node* node, Value* curToken) {
+    if (node->kind() == prim::MemoryFence) {
```
Once we insert a memory fence, the next call to `visitNode` visits that memory fence node (since it gets added to the block's node list). This check is to avoid an infinite loop; I'll add a comment to clarify.

suo committed:
Address cosmetic comments

suo committed:
Distinguish between graph->insert and compiler schema matching

Allow for different schema matching strategies when emitting builtins,
so that graph->insert() will not construct a dummy world token

suo committed:
Improvements to AnnotateEffects

- Const correctness
- isMutableType
- Actually check that we're fencing using a mutable value
- More comments

suo committed:
special world token nodes for graphs, fix function inlining

suo committed:
clean up unnecessary constant prop changes

michaelsuo commented:
Okay, comments should be addressed. I'm waiting on #10812 to properly test function inlining though, since right now we can't pass lists as arguments.

ezyang commented:
Are we going to eventually distinguish between mutable lists and immutable lists in the type system? I would quite like that, and it will help us do better when doing shape inference (since immutable lists don't have to deal with the covariance/contravariance problem)

suo committed:
fix build

suo committed:
add simple test for function inlining

zdevito commented:
I think we should eventually add a `is_mutable` flag to ListType and DynamicType. That would help distinguish mutable for simple cases.  A mutability inference pass can run after the compiler but before  inserting fencing that would performance this inference. That way the fencing pass will easily know when to add fences and when not to. This doesn't have to be in this PR.

suo committed:
Clean up after rebase

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -535,6 +536,44 @@ Value* tryCreateList(
+  std::vector<NamedValue> modifiedInputs(inputs.begin(), inputs.end());
+  if (schema.is_mutable) {
+    // Add a dummy world token to be matched against
+    const auto worldToken = graph.insertConstant(World());
```
A `prim::Constant` world and `prim::EntryWorld` would have the same behavior the in annotate effects pass. If we just inserted a node `prim::LoadCurrentWorld` in both cases you can unify the concept.

michaelsuo replied:
Just to clarify about naming: `prim::Store` takes stack values and puts them to registers, returning them as output. `prim::Load` takes the registers and puts them on the stack. 

So in this case, prim::EntryWorld is more like prim::Store, right? Since it is taking something from the outside world and spawning it on the stack.

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -687,20 +721,27 @@ Value* emitBuiltinCall(
-              op, failure_messages, loc, graph, name, inputs, attributes,
-              convert_tensors_to_nums)) {
-        return result;
+      const auto matched_inputs = matchStrategy(
```
This is huge code smell that we got something wrong here. We now have `Impl` variants of builtin and strategy patterns. All to just get over the fact that there are now two ways of matching schema! There should only be a single way to emit ops and check their schema is valid.

Here is what I think we should do: just make emitBuiltinCall universally insert the world token if the schema is mutable. It should emit a `prim::LoadWorld` op.  `prim::LoadWorld` should just fault on use. It will then be a passes responsibility if they are emitting a mutable op to fix up the prim::LoadWorld nodes. This kicks the can down the road but I think it is better than complicated the `emit` logic in this way. We can improve this later. For instance, we can add a `std::function<Value*()> world_provider = nullptr` to emit that allows you to specify how to generate a World input if needed.

zdevito commented on torch/csrc/jit/ir.cpp:
```diff
@@ -516,6 +529,15 @@ void Block::cloneFrom(Block * src, std::function<Value*(Value*)> value_map) {
     graph->setStage(std::max(graph->stage(), input->stage()));
   }
+  const auto entryWorld = src->entryWorld();
+  if (!entryWorld->outputs().empty()) {
```
These conditionals mean that this code gets tested very rarely. One of the reasons I want it non-conditional is that we find bugs in how cloning and other code interacts with the world token handling early.

zdevito commented on torch/csrc/jit/passes/annotate_effects.cpp:
```diff
@@ -0,0 +1,313 @@
+      //
+      // We can ignore all nodes in between the entry and exit tokens, since
+      // tokens have already been threaded through then.
+      bool skip = false;
```
You can tell if it should be a no-op if the op itself is already fenced (i.e. all inputs/outputs are coming from/to memory fences).

zdevito reviewed (commented):
I have a bunch of suggestions for how we can make the handling of world tokens across functions and in emitBuiltinCall clearer and more concise. I think it is important we get this right because we are going to be building a lot on top of this code.

apaszke reviewed (commented):
Not a complete review. Some initial comments that I don't want to get lost.

michaelsuo commented on torch/csrc/jit/ir.cpp:
```diff
@@ -317,18 +317,24 @@ void Node::lint() const {
     JIT_ASSERT(inputs_.size() == 0);
+  IR_ELSEIF(EntryWorld)
+    // EntryWorld has 0 or 1 inputs/outputs
+    JIT_ASSERT(inputs_.size() <= 1);
```
yeah this is wrong, EntryWorld should only have outputs and ExitWorld should only have inputs, just like graph input/output

michaelsuo commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1710,6 +1799,14 @@ std::vector<Value*> inlineCallTo(Graph& g, Graph& callee, ArrayRef<Value*> input
+    auto* newWorld = g.insertNode(g.createClone(exitWorld, value_map_func));
+    for (size_t i = 0; i < exitWorld->outputs().size(); ++i) {
+      value_map[exitWorld->outputs()[i]] = newWorld->outputs()[i];
+    }
```
The two world token threads are separate until the AnnotateEffects pass joins them.

suo committed:
- Add a LoadWorld and StoreWorld op to distinguish between special nodes
and concrete nodes
- Fix up Entry/Exit inputs and outputs to actually make sense

suo committed:
Emit dummy node always instead of forking tryMatchSchema

suo committed:
making inlineCallTo make a copy internally

michaelsuo commented on torch/csrc/jit/passes/annotate_effects.cpp:
```diff
@@ -0,0 +1,313 @@
+    // Handle inlined functions. Inlined functions will expose their Entry and
+    // Exit tokens as regular nodes. These exposed nodes provide fixed points
+    // to thread the current world token through.
+    if (node->kind() == prim::EntryWorld) {
```
this is unnecessary, removing

suo committed:
fence off all inputs and outputs

suo committed:
Annotate inlined nodes normally instead of skipping

michaelsuo commented:
Okay, all comments are addressed, except: should we move `EntryWorld` and `ExitWorld` out of the top-level block and into the graph? This is not consistent with how graph inputs/outputs are done, but maybe we want to start moving special nodes out of blocks. It also violates the invariant that every node belongs to exactly one block, but I'm not sure how important that is given the specialness of these nodes.

Also, don't mind the test failures; one unfortunate side effect of adding an output value to `EntryWorld` no matter what is that it changes the value names in every graph. I wanted to get feedback on the changes before fixing up ~80 tests 😛.

michaelsuo commented:
One other thing I realized: we don't actually need an `ExitWorld` node, since the AnnotateEffects pass infers the last used token at all times anyway. I plan to remove it unless people want to keep around for some future purpose (do we ever call entire graphs from other graphs without inlining?).

suo committed:
cleanup to AnnotateEffects

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1699,23 +1741,41 @@ std::shared_ptr<SugaredValue> SimpleValue::attr(SourceRange loc, Method & m, con
+  // TODO(suo): a cleaner way to handle this might be to have inlined graphs
+  // be a special prim::Inlined node, so that we can clone the entire block's
+  // metadata and don't need to expose EntryWorld and ExitWorld like this.
+  auto* storeWorld = g.insertNode(g.create(prim::StoreWorld));
```
The words Store and Load and store are flipped here.  A load should have an output. A store should have an input.

michaelsuo replied:
I'm a little confused by the terms: right now, graphs start with a `prim::Store` (converted from a `prim::Param`) and end with a `prim::Load` (converted from `prim::Return`), so I was just following that convention. Should I change that too? (I agree that your suggested meaning makes more sense to me)

zdevito replied:
Yeah, those instructions are flipped as well.

zdevito commented on torch/csrc/jit/interned_strings.h:
```diff
@@ -56,6 +56,13 @@ namespace torch { namespace jit {
+  _(prim, ExitWorld)               \
+  _(prim, LoadWorld)               \
+  _(prim, StoreWorld)              \
+  _(prim, DummyWorld)              \
```
prim::DummyWorld and prim::LoadWorld do the same thing: they load the 'current' world token, for whatever definition of current there is.

michaelsuo replied:
Hm, the main reason for having `DummyWorld` is that the program will fault at runtime if that instruction is every executed by the interpreter.

Whereas every program should start with a `LoadWorld` instruction to set up the world token. This has the effect of forcing pass writers to deal with any `DummyWorld`s they are emitting.

Is there a better way to distinguish these two use-cases with a single op?

zdevito replied:
I see, that makes sense as a reason for having two ops.

suo committed:
throw a more reasonable error on DummyWorld

jamesr66a commented on torch/csrc/jit/passes/to_batch.cpp:
```diff
@@ -525,11 +526,11 @@ void ToBatch::toBatch(Block* block, Block* res_block) {
   to_batch.toBatch(graph->block(), res_graph->block());
-  // std::cout<<res_graph->toString()<<std::endl;
+
+  EliminateDeadCode(res_graph);
```
What happened here?

michaelsuo replied:
The inlining that this pass does introduced extraneous world tokens. They are harmless affected the readability of the graph, so I added a DCE pass to clean things up.

suo committed:
swap `load` and `store` in the interpreter

suo committed:
fixes after rebase

zdevito commented on torch/csrc/jit/ir.h:
```diff
@@ -858,6 +861,8 @@ struct Block {
   // having corner cases where the list is empty.
   Node * const output_;
   Node * const input_;
+  Node * const entry_world_;
```
Are entry_world_ and exit_world_ used at the block level or just at the graph level?  If they are only used at the graph level, they should be defined on Graph and not Block.  Someone might call `entryWorld()` on the block, expecting it to be the loop-carried value that is the world token that is the input to a loop block body, but instead get an unused Value* instead.

zdevito commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -615,18 +651,24 @@ RegisterOperators reg2({
     Operator("aten::eq(float[] a, float[] b) -> int", listEq<Shared<DoubleList>>),
     Operator("aten::eq(Tensor[] a, Tensor[] b) -> int", listEq<Shared<TensorList>>),
 
+    // Append `el` to `list`
```
Can you add this to CREATE_LIST_OPS instead well? Otherwise you cannot append to a list of lists because it doesn't implement list of lists.

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -636,35 +638,42 @@ at::optional<MatchedSchema> tryMatchSchema(
+    at::ArrayRef<NamedValue> attributes,
+    std::ostream& failure_messages,
+    bool convert_tensors_to_nums) {
+  std::vector<NamedValue> modifiedInputs(inputs.begin(), inputs.end());
```
Can you just put these lines at the beginning of tryMatchSchema. There are like 20 lines of parameter passing to the Impl version to add these 5 lines of code here.

michaelsuo replied:
oops, this function used to be more complicated I think. Will remove

zdevito commented on torch/csrc/jit/type.h:
```diff
@@ -370,6 +371,32 @@ struct TORCH_API CompleteTensorType : public TensorType {
 
+// This type is a token used to represent effectful computation in the IR.
+// See the AnnotateEffects pass for how it is used.
+struct WorldType;
```
Do a search for `GeneratorType` in the codebase. There are a few switch statements that are missing clauses to handle the WorldType (such as module import/export).

zdevito reviewed (approved):
This looks good! I have minor comments that should be easy to address. You touched an ONNX file so ask @jamesr66a to review/stamp the ONNX changes.

michaelsuo commented on test/expect/TestBatched.test_for.expect:
```diff
@@ -4,19 +4,19 @@ graph(%x.1_data : Dynamic
-      %19 : int = prim::Constant[value=1]()
-      %data : Dynamic = aten::where(%mask, %data.1, %5_data)
-      -> (%19, %data, %mask, %dims)
+  %7 : int = prim::Constant[value=10]()
```
It's the effect of having a `entryWorld` value in every graph. In order to fix it, we'd need to distinguish "control" values and put their unique ids in a different namespace. I don't think it's worth the extra bookkeeping.

suo committed:
review comments

michaelsuo commented on test/test_jit.py:
```diff
@@ -1647,7 +1647,7 @@ def single_if(a, b):
         script_if = torch.jit.script(single_if)
         graph = torch.to_batch_graph(script_if.graph)
-        self.assertExpected(str(graph))
+        self.assertExpected(canonical(graph))
```
this ended up being weird (changing TestBatch to inherit from JitTestCase caused some tests to fail). I'll investigate and fix in a difference PR.
